src/main.rs

```diff
@@ -282,58 +280,72 @@ fn read_patterns(file: &PathBuf) -> Result<Vec<String>> {
     Ok(v)
 }
 
+/// Runs fqgrep and converts Option<u8> to ExitCode
+///
+/// Set the exit code:
+/// - exit code SUCCESS if there were matches
+/// - exit code 1 if there were no matches
+/// - exit code 2 if fqgrep returned an error
+/// - exit code 101 if panic
 fn main() -> ExitCode {
-    // Set the exit code:
-    // - exit code 0 if there were matches
-    // - exit code 1 if there were no matches
-    // - exit code 2 if fqgrep returned an error
-    // - exit code 101 if fqgrep panicked
-    let outer = std::panic::catch_unwind(fqgrep);
+    // Receives u8 from
+    if let Some(fqgrep_output) = fqgrep(&setup()) {
+        ExitCode::from(fqgrep_output)
+    } else {
+        ExitCode::SUCCESS
+    }
+}
+
+/// Runs fqgrep_from_opts and returns None upon success and an error number if error or zero matches
+///
+/// - None if there were matches
+/// - 1 if there were no matches
+/// - 2 if fqgrep_from_opts returned an error
+/// - 101 if fqgrep_from_opts panicked
+fn fqgrep(opts: &Opts) -> Option<u8> {
+    let outer = std::panic::catch_unwind(|| fqgrep_from_opts(opts));
     match outer {
         Err(_) => {
             eprintln!("Error: fqgrep panicked.  Please report this as a bug!");
-            ExitCode::from(101)
+            Some(101)
         }
         Ok(inner) => match inner {
-            Ok(0) => ExitCode::from(1),
-            Ok(_) => ExitCode::SUCCESS,
+            Ok(0) => Some(1),
+            Ok(_) => None,
             Err(e) => {
                 eprintln!("Error: {}", e);
-                ExitCode::from(2)
+                Some(2)
             }
         },
     }
 }
 
-fn fqgrep() -> Result<usize> {
-    let mut opts = setup();
-    fqgrep_from_opts(&mut opts)
-}
-
 #[allow(clippy::too_many_lines)]
-fn fqgrep_from_opts(opts: &mut Opts) -> Result<usize> {
+fn fqgrep_from_opts(opts: &Opts) -> Result<usize> {
+    let mut mutable_opts = opts.clone();
```

Comment:
Can we just use `opts` here so that the previous one goes out of scope?  Then there's no danger later in using `opts` versus `mutable_opts`, which there is now.
```suggestion
    let mut opts = opts.clone();
```